platform/linux-generic/include/odp_ring_st_internal.h

```diff
@@ -0,0 +1,118 @@
+/* Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#ifndef ODP_RING_ST_INTERNAL_H_
+#define ODP_RING_ST_INTERNAL_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/hints.h>
+#include <odp_align_internal.h>
+
+/* Basic ring for single thread usage. Operations must be synchronized by using
+ * locks (or other means), when multiple threads use the same ring. */
+typedef struct {
+	uint32_t head;
+	uint32_t tail;
+	uint32_t mask;
+	uint32_t *data;
+
+} ring_st_t;
+
+/* Initialize ring. Ring size must be a power of two. */
+static inline void ring_st_init(ring_st_t *ring, uint32_t *data, uint32_t size)
+{
+	ring->head = 0;
+	ring->tail = 0;
+	ring->mask = size - 1;
+	ring->data = data;
+}
+
+/* Dequeue data from the ring head. Max_num is smaller than ring size.*/
+static inline uint32_t ring_st_deq_multi(ring_st_t *ring, uint32_t data[],
+					 uint32_t max_num)
+{
+	uint32_t head, tail, mask, idx;
+	uint32_t num, i;
+
+	head = ring->head;
+	tail = ring->tail;
+	mask = ring->mask;
+	num  = tail - head;
+
+	/* Empty */
+	if (num == 0)
+		return 0;
+
+	if (num > max_num)
+		num = max_num;
+
+	idx = head & mask;
+
+	for (i = 0; i < num; i++) {
+		data[i] = ring->data[idx];
+		idx     = (idx + 1) & mask;
+	}
+
+	ring->head = head + num;
+
+	return num;
+}
+
+/* Enqueue data into the ring tail. Num_data is smaller than ring size. */
+static inline uint32_t ring_st_enq_multi(ring_st_t *ring, const uint32_t data[],
+					 uint32_t num_data)
+{
+	uint32_t head, tail, mask, size, idx;
+	uint32_t num, i;
+
+	head = ring->head;
+	tail = ring->tail;
+	mask = ring->mask;
+	size = mask + 1;
+	num  = size - (tail - head);
```

Comment:
Tail and head indexes are (masked from) uint32_t and do not wrap around when the ring is full. I think you assume that the store index is 0...size-1, while it's full uint32_t which is then masked to get the actual index.

For example:
size = 100;

Empty:
head = 100
tail = 100
num = 100 - 100 = 0

Full:
head = 100
tail = 200
num = 200 - 100 = 100

Wrap uint32_t + full:
head = 0xFFFFFF9C
tail = 0
num = 0 - 0xFFFFFF9C = 0x64 = 100

So, no abs() needed. Ring size can be 4096, instead of 4095.

Reply:
No since ring size will be much smaller than 4 billion.

Reply:
Please see above discussion. The comment stands.